File: src/main/java/cube8540/oauth/authentication/credentials/authority/application/DefaultSecurityMetadataLoadService.java
package cube8540.oauth.authentication.credentials.authority.application;

import cube8540.oauth.authentication.credentials.authority.domain.AuthorityCode;
import cube8540.oauth.authentication.credentials.authority.domain.ResourceMethod;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResource;
import cube8540.oauth.authentication.credentials.authority.domain.SecuredResourceRepository;
import cube8540.oauth.authentication.credentials.authority.security.SecurityMetadataLoadService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.ConfigAttribute;
import org.springframework.security.access.SecurityConfig;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
import org.springframework.security.web.util.matcher.RequestMatcher;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class DefaultSecurityMetadataLoadService implements SecurityMetadataLoadService {

    private final SecuredResourceRepository repository;

    @Autowired
    public DefaultSecurityMetadataLoadService(SecuredResourceRepository repository) {
        this.repository = repository;
    }

    @Override
    public Map<RequestMatcher, Collection<ConfigAttribute>> loadSecurityMetadata() {
        return repository.findAll().stream().collect(Collectors.toMap(this::requestMatcher, resource -> authorityToConfigAttribute(resource.getAuthorities())));
    }

    private RequestMatcher requestMatcher(SecuredResource securedResource) {
        if (securedResource.getMethod().equals(ResourceMethod.ALL)) {
            return new AntPathRequestMatcher(securedResource.getResource().toString());
        } else {
            return new AntPathRequestMatcher(securedResource.getResource().toString(), securedResource.getMethod().toString());
        }
    }

    private Collection<ConfigAttribute> authorityToConfigAttribute(Set<AuthorityCode> authorities) {
        return Optional.ofNullable(authorities).orElse(Collections.emptySet())
                .stream().map(AuthorityCode::getValue).map(SecurityConfig::new).collect(Collectors.toList());
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/resource/infra/DefaultSecurityMetadataLoadService.kt
package cube8540.oauth.authentication.resource.infra

import cube8540.oauth.authentication.resource.domain.AccessibleAuthority
import cube8540.oauth.authentication.resource.domain.ResourceMethod
import cube8540.oauth.authentication.resource.domain.SecuredResource
import cube8540.oauth.authentication.resource.domain.SecuredResourceRepository
import cube8540.oauth.authentication.security.ScopeSecurityConfig
import cube8540.oauth.authentication.security.SecurityMetadataLoadService
import org.springframework.security.access.ConfigAttribute
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import org.springframework.security.web.util.matcher.RequestMatcher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class DefaultSecurityMetadataLoadService(private val repository: SecuredResourceRepository):
    SecurityMetadataLoadService {

    @Transactional
    override fun loadSecurityMetadata(): Map<RequestMatcher, Collection<ConfigAttribute>> =
        repository.findAll().associate { requestMatcher(it) to authorityToConfigAttribute(it.authorities) }

    private fun requestMatcher(securedResource: SecuredResource): RequestMatcher = when (securedResource.method) {
        ResourceMethod.ALL -> {
            AntPathRequestMatcher(securedResource.resource.toString())
        }
        else -> {
            AntPathRequestMatcher(securedResource.resource.toString(), securedResource.method.toString())
        }
    }

    private fun authorityToConfigAttribute(authorities: Set<AccessibleAuthority>?): Collection<ConfigAttribute> {
        val configAttributes = HashSet<ConfigAttribute>()

        authorities?.forEach { configAttributes.add(ScopeSecurityConfig(it.authority)) }

        return configAttributes
    }
}